agent/pkg/docker/docker.go

```diff
@@ -0,0 +1,464 @@
+	AgentLabel = "ai.determined.container.agent"
+	// ClusterLabel gives the cluster the container is managed by.
+	ClusterLabel = "ai.determined.container.cluster"
+	// MasterLabel gives the master the container is managed by.
+	MasterLabel = "ai.determined.container.master"
+
+	// ImagePullStatsKind describes the IMAGEPULL event.
+	ImagePullStatsKind = "IMAGEPULL"
+)
+
+type (
+	// ContainerWaiter contains channels to wait on the termination of a running container.
+	// Results on the Waiter channel indicate changes in container state, while results on the
+	// Errs channel indicate failures to watch for updates.
+	ContainerWaiter struct {
+		Waiter <-chan dcontainer.ContainerWaitOKBody
+		Errs   <-chan error
+	}
+	// Container contains details about a running container and waiters to await its termination.
+	Container struct {
+		ContainerInfo   types.ContainerJSON
+		ContainerWaiter ContainerWaiter
+	}
+)
+
+// Client wraps the Docker client, augmenting it with a few higher level convenience APIs.
+type Client struct {
+	// Configuration details. Set during initialization, never modified afterwards.
+	credentialStores map[string]*credentialStore
+	authConfigs      map[string]types.AuthConfig
+
+	// System dependencies. Also set during initialization, never modified afterwards.
+	cl     *client.Client
+	logger *logrus.Entry
+}
+
+// NewClient populates credentials from the Daemon config returns a new Client that uses them.
+func NewClient(cl *client.Client) *Client {
+	d := &Client{
+		cl:     cl,
+		logger: logrus.WithField("component", "docker-client"),
+	}
+
+	stores, auths, err := processDockerConfig()
+	if err != nil {
+		d.logger.Infof("couldn't process ~/.docker/config.json %v", err)
+	}
+	if len(stores) == 0 {
+		d.logger.Info("can't find any docker credential stores, continuing without them")
+	}
+	if len(auths) == 0 {
+		d.logger.Info("can't find any auths in ~/.docker/config.json, continuing without them")
+	}
+	d.credentialStores, d.authConfigs = stores, auths
+
+	return d
+}
+
+// Inner returns the underlying Docker client, to be used sparingly.
+// TODO(???): Consolidate around usage of the wrapper client, remove this.
+func (d *Client) Inner() *client.Client {
+	return d.cl
+}
+
+// Close closes the underlying Docker client.
+func (d *Client) Close() error {
+	return d.cl.Close()
+}
+
+// ReattachContainer looks for a single running or terminated container that matches the given
+// filters and returns whether it can be reattached or has terminated.
+func (d *Client) ReattachContainer(
+	ctx context.Context,
+	filter filters.Args,
+) (*Container, *aproto.ExitCode, error) {
+	containers, err := d.cl.ContainerList(ctx, types.ContainerListOptions{Filters: filter})
+	if err != nil {
+		return nil, nil, fmt.Errorf("while reattaching container: %w", err)
+	}
+
+	if len(containers) > 1 {
+		return nil, nil, errors.New("reattach filters matched more than one container")
+	}
+
+	context.Background()
+
```

Comment:
🤔 
```suggestion
```

Reply:
resolving this that github failed to auto-resolve when i accepted the suggestion.